src/java/org/apache/cassandra/service/cleanupstate/CleanupStateTracker.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service.cleanupstate;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Map;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+
+public class CleanupStateTracker
+{
+    @VisibleForTesting
+    static final Instant MIN_TS = Instant.EPOCH;
+    private static final String CLEANUP_STATE_FILE_NAME = "node_cleanup_state.json";
+
+    private final CleanupState state;
+    private final CleanupStatePersister persister;
+
+    public CleanupStateTracker() throws IOException
+    {
+        this.persister = new CleanupStatePersister(cleanupStateFileLocation());
+        this.state = new CleanupState(persister.readCleanupStateFromFile());
```

Comment:
What do you think about making CleanupStatePersister a property of CleanupState? So when you update CleanupState the file is updated in the same call

Reply:
discussed offline: not a strong preference so OK to stay as is